Makefile

```diff
@@ -1,21 +1,91 @@
-RETOOL=$(CURDIR)/_tools/bin/retool
-PATH := ${PWD}/bin:${PWD}/ENV/bin:${PATH}
+SHELL := /bin/bash -o pipefail
+UNAME_OS := $(shell uname -s)
+UNAME_ARCH := $(shell uname -m)
+
+TMP_BASE := .tmp
+TMP := $(TMP_BASE)/$(UNAME_OS)/$(UNAME_ARCH)
+TMP_BIN = $(TMP)/bin
+
+DEP_VERSION := 0.5.0
```

Comment:
Minor complaint, this means that if we move to a newer DEP version cruft will start to build up with multiple versions of DEP in the folder. This is true for all the apps.

Reply:
They won't - I purposefully write them to the same location `$TMP_BIN/dep`. You can put the version as part of the binary name, which I have done in the past, but that means you can't call `dep`, instead you have to call `$(DEP)`, and for example with `protoc`, this is problematic with how the scripts are set up (the scripts invoke `protoc`, not `protoc-3.7.0`, since they don't have access to this version information).

If you upgrade the tools, it is good to call `make clean`.

Reply:
I missed that, but I think that this is adding unknown tribal knowledge complexity to someone working on the project. I would advocate at least a section in the readme about this. Essentially this is rolling out a mini package manager without all the hard parts that make writing a package manager hard.

For example if I work on twirp and pull a new copy of master I now need to verify myself if the makefile changed a version, and if it did I have to clean `.tmp` myself.

Reply:
Just my opinion: having just gone through building Twirp the first time, this system should already be **way** easier than the system using retool - I had to fight a bunch of things to get deps updated. You're right in that it does it's own package management, and the simplicity of it is the killer feature - nothing to learn, it's basically hermetic (outside of git, go, curl, make), and it works easily on both Darwin and Linux.

I can add something to CONTRIBUTING.md if you think it's important.

Reply:
One other note: This is no different than the current setup in this manner. When I check out a new copy of master right now, I need to run `make setup` to verify that `retool build` is called (or I need to know to call `retool`). `setup` isn't called before every relevant target.

Reply:
I've now fixed the problem - each tool now depends on a file `$(TMP)/versions/NAME/VERSION` instead of the actual binary. When installing, the `NAME` directory is deleted, and the file is created with `touch`. This makes it so that if versions switch, you will be updated, including if you switch back and forth, but the binary will still have the same name.

I've verified this works locally as expected by switching versions back and forth, and making sure that the install happens when a version is changed, but not at other times.

Reply:
> One other note: This is no different than the current setup in this manner. When I check out a new copy of master right now, I need to run make setup to verify that retool build is called (or I need to know to call retool). setup isn't called before every relevant target.

To be clear, `setup` doesn't need to be called every time. Retool checks whether tools are synced to the right version on [every `retool do` invocation](https://github.com/twitchtv/retool/blob/master/main.go#L52-L54), so they're always at the right version. But yes, the touched filenames bring this in to your Makefile version, too.